### PR TITLE
update NEP3 class to consider NEP4

### PR DESCRIPTION
**Pull request by brucefan1983**
See the relevant PR in GPUMD: https://github.com/brucefan1983/GPUMD/pull/277

### CODE REVIEW

File: src/nep.cpp [hunk 1]
```diff
@@ -1243,7 +1243,7 @@ void find_descriptor_for_lammps(
 
     double F = 0.0, Fp[MAX_DIM] = {0.0}, latent_space[MAX_NEURON] = {0.0};
     apply_ann_one_layer(
-      annmb.dim, annmb.num_neurons1, annmb.w0, annmb.b0, annmb.w1, annmb.b1, q, F, Fp,
+      annmb.dim, annmb.num_neurons1, annmb.w0[t1], annmb.b0[t1], annmb.w1[t1], annmb.b1, q, F, Fp,
```
Comment (elindgren) [inline src/nep.cpp:1246]: Why is the variable name for the different types `t1` instead of, e.g., `type`?

File: src/nep.h [hunk 2]
```diff
@@ -43,9 +43,9 @@ class NEP3
     int dim = 0;
     int num_neurons1 = 0;
     int num_para = 0;
-    const double* w0;
-    const double* b0;
-    const double* w1;
+    const double* w0[100];
+    const double* b0[100];
+    const double* w1[100];
```
Comment (elindgren) [inline src/nep.h:48]: So there can be no more than 100 different atom types, and thus a maximum of a hundred different networks?

**Review by elindgren — APPROVED** [review]
:racehorse:

File: src/nep.h (hunk 2 above)
Comment (brucefan1983) [reply src/nep.h:48]: yes, 100 is just my estimate for the maximum number of elements in the potential :-)

File: src/nep.cpp (hunk 1 above)
Comment (brucefan1983) [reply src/nep.cpp:1246]: just to make it shorter :-)
In the code:
`t1`  = type of atom `n1`
`t2` = type of aotm `n2`

Comment (brucefan1983) [reply src/nep.cpp:1246]: It was defined here:
```
int t1 = g_type[n1] - 1; // from LAMMPS to NEP convention
```

File: src/nep.h (hunk 2 above)
Comment (elindgren) [reply src/nep.h:48]: Sounds reasonable, if someone tries to run a simulation with more than 100 species they will have other problems :rofl:

File: src/nep.cpp (hunk 1 above)
Comment (elindgren) [reply src/nep.cpp:1246]: Alright :+1:

**Review by bigd4 — APPROVED** [review]
0.0

**Pull request merged**